File: LineSDKStarterObjC/LineSDKStarterObjC/ViewController.m
//
//  ViewController.m
//  LineSDKStarterObjC
//
//  Copyright (c) 2015 Line corp. All rights reserved.
//

#import "ViewController.h"
#import <LineAdapter/LineAdapter.h>
#import <LineAdapterUI/LineAdapterUI.h>

@interface ViewController ()

@end

@implementation ViewController {
    LineAdapter *adapter;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    adapter = [LineAdapter adapterWithConfigFile];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(authorizationDidChange:) name:LineAdapterAuthorizationDidChangeNotification object:nil];
}

- (IBAction)loginWithLine:(id)sender {
    if (adapter.isAuthorized) {
        [self alert:@"Already authorized" message:nil];
        return;
    }

    if (!adapter.canAuthorizeUsingLineApp) {
        [self alert:@"LINE is not installed" message:nil];
        return;
    }

    [adapter authorize];
}

- (IBAction)loginInApp:(id)sender {
    if (adapter.isAuthorized) {
        [self alert:@"Already authorized" message:nil];
        return;
    }
    LineAdapterWebViewController *viewController = [[LineAdapterWebViewController alloc] initWithAdapter:adapter withWebViewOrientation:kOrientationAll];
    viewController.navigationItem.leftBarButtonItem = [LineAdapterNavigationController barButtonItemWithTitle:@"Cancel" target:self action:@selector(cancel:)];
    LineAdapterNavigationController *navigationController = [[LineAdapterNavigationController alloc] initWithRootViewController:viewController];
    [self presentViewController:navigationController animated:YES completion:nil];

}

- (IBAction)tryApi:(id)sender {
    if (!adapter.isAuthorized) {
        [self alert:@"Login first!" message:nil];
        return;
    }

    [adapter.getLineApiClient getMyProfileWithResultBlock:^(NSDictionary *aResult, NSError *aError) {
        if (aError) {
            [self alert:@"Error occured!" message: aError.localizedDescription];
            return;
        }

        NSString *displayName = aResult[@"displayName"];
        [self alert:[NSString stringWithFormat:@"Your name is %@", displayName] message:nil];
    }];
}

- (IBAction)logout:(id)sender {
    [adapter unauthorize];
    [self alert:@"Logged out" message:nil];
}

- (void)alert:(NSString *)title message:(NSString *)message {
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:title message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [alert show];
}

- (void)cancel:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Handle Notification

- (void)authorizationDidChange:(NSNotification *)notification {
    LineAdapter *lineAdapter = notification.object;
    if (lineAdapter.isAuthorized) {
        [self alert:@"Login success!" message:nil];
        [self dismissViewControllerAnimated:YES completion:nil];
        return;
    }

    NSError *error = notification.userInfo[@"error"];
    if (error) {
        [self alert:@"Login error!" message:error.localizedDescription];
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}


@end

File: LineSDKStarterSwift/LineSDKStarterSwift/ViewController.swift
//
//  ViewController.swift
//  LineSDKStarterSwift
//
//  Copyright (c) 2015 Line corp. All rights reserved.
//

import UIKit

final class ViewController: UIViewController {
    private let adapter = LineAdapter.default()!;

    override func viewDidLoad() {
        super.viewDidLoad()
        startObserveLineAdapterNotification()
    }

    func startObserveLineAdapterNotification() {
        NotificationCenter.default.addObserver(self, selector: #selector(ViewController.authorizationDidChange(_:)),
                                               name: NSNotification.Name.LineAdapterAuthorizationDidChange, object: nil)
    }

    @IBAction func loginWithLine(_ sender: AnyObject) {
        if adapter.isAuthorized {
            alert("Already authorized", message: "")
            return
        }

        if !adapter.canAuthorizeUsingLineApp {
            alert("LINE is not installed", message: "")
            return
        }
        adapter.authorize()
    }

    @IBAction func loginInApp(_ sender: AnyObject) {
        if adapter.isAuthorized {
            alert("Already authorized", message: "")
            return
        }

        let viewController = LineAdapterWebViewController(adapter: adapter, with: LineAdapterWebViewOrientation.all)
        viewController.navigationItem.leftBarButtonItem = LineAdapterNavigationController.barButtonItem(withTitle: "Cancel", target: self, action: #selector(ViewController.cancel(_:)))
        let navigationController = LineAdapterNavigationController(rootViewController: viewController)
        present(navigationController, animated: true, completion: nil)
    }

    @IBAction func tryApi(_ sender: AnyObject) {
        if !adapter.isAuthorized {
            alert("Login first!", message: "")
            return
        }

        adapter.getLineApiClient().getMyProfile {[unowned self] (profile, error) -> Void in
            if let error = error {
                self.alert("Error occured!", message: error.localizedDescription)
                return
            }

            let displayName = profile?["displayName"] as! String
            self.alert("Your name is \(displayName)", message: "")
        }
    }

    @IBAction func logout(_ sender: AnyObject) {
        adapter.unauthorize()
        alert("Logged out", message: "")
    }

    fileprivate func alert(_ title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
        let buttonAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.cancel)
        alert.addAction(buttonAction)
        present(alert, animated: true, completion: nil)
    }
}

extension ViewController {

    func cancel(_ sender: AnyObject) {
        dismiss(animated: true, completion: nil)
    }

    func authorizationDidChange(_ notification: Notification) {
        let adapter = notification.object as! LineAdapter

        if adapter.isAuthorized {
            dismiss(animated: true, completion: nil)
            alert("Login success!", message: "")
            return
        }

        if let error = notification.userInfo?["error"] as? NSError {
            dismiss(animated: true, completion: nil)
            alert("Login error!", message: error.localizedDescription)
        }

    }

}
